include/xtensor/xstrided_view_base.hpp

```diff
@@ -174,8 +174,11 @@ namespace xt
             using reference = typename xexpression_type::reference;
             using const_reference = typename xexpression_type::const_reference;
 
-            using iterator = typename xexpression_type::iterator;
-            using const_iterator = typename xexpression_type::const_iterator;
+            // using iterator = typename xexpression_type::layout_iterator<XTENSOR_DEFAULT_LAYOUT>;
+            // using const_iterator = typename xexpression_type::const_layout_iterator<XTENSOR_DEFAULT_LAYOUT>;
+
+            using iterator = decltype(std::declval<std::decay_t<CT>>().template begin<XTENSOR_DEFAULT_LAYOUT>());
+            using const_iterator = decltype(std::declval<std::decay_t<const CT>>().template begin<XTENSOR_DEFAULT_LAYOUT>());
```

Comment:
1. Using `decltype(std::declval)` but we should really be able to use the typedefs in the comments above.

   @JohanMabille Should we declare the layout iterator types in all expression types?

2. At the moment, `flat_expression_adaptor` must use the default layout and not the traversal layout, hance the change.

Reply:
Yes each expression should provide `layout_iterator` and `const_layout_iterator`, either by publicly inheriting from `xconst_iterable` or `xiterable`, or by defining them.

Reply:
Does it work with using `const_iterator = typename xexpression_type::template const_layout_iterator<XTENSOR_DEFAULT_LAYOUT>;` ?

Reply:
However, `begin`  and `end` do not necessarily return `layout_iterator`; in the case of `xcontiguously_iterable`, they return a `storage_iterator` if the iteration order and the layout of the container are the same.

Reply:
Also the second type should be:
```cpp
using const_iterator = decltype(std::declval<const std::decay_t<CT>>().template begin<XTENSOR_DEFAULT_LAYOUT>());
```
Otherwise the `const` is removed by `std::decay_t`.

Reply:
Or even better:
```cpp
using const_iterator = decltype(std::declval<std::decay_t<CT>>().template cbegin<XTENSOR_DEFAULT_LAYOUT>());
```